### PR TITLE
Rewrite the async code in `BasicQueue`

### PR DESCRIPTION
**Pull request by bkchr**
This is some smaller change to rewrite the async code in `BasicQueue`. I
require this for some other pr I'm working on ;)

### CODE REVIEW

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 1]
```diff
@@ -405,22 +361,51 @@ fn import_many_blocks<B: BlockT, V: Verifier<B>, Transaction>(
...
-			waiting = Some(Delay::new(delay_between_blocks));
+		if delay_between_blocks != Duration::default() && !has_error {
+			Delay::new(delay_between_blocks).await;
+		} else {
+			Yield::new().await
+		}
+	}
+}
+
+/// A future that will `yield` on the first call of `poll`.
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:385]: which 
```suggestion
/// A future that will always `yield` on the first call of `poll` but schedules the current task for
/// re-execution.

```

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 2]
```diff
@@ -405,22 +361,51 @@ fn import_many_blocks<B: BlockT, V: Verifier<B>, Transaction>(
...
 			has_error = true;
 		}
 
 		results.push((import_result, block_hash));
 
-		// Notifies the current task again so that we re-execute this closure again for the next
-		// block.
-		if delay_between_blocks != Duration::new(0, 0) {
-			waiting = Some(Delay::new(delay_between_blocks));
+		if delay_between_blocks != Duration::default() && !has_error {
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:377]: This `!has_error` is a behavior change, but it makes sense, there's no point in delaying cancellations 👍 (in practice we don't use this delay anyway).

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 3]
```diff
@@ -307,29 +291,27 @@ impl<B: BlockT, Transaction: Send> BlockImportWorker<B, Transaction> {
 	}
 }
 
+/// Result of [`import_many_blocks`].
+struct ImportManyBlocksResult<B: BlockT> {
+	/// The number of imported blocks successfully.
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:296]:
```suggestion
	/// The number of blocks imported successfully.
```

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 4]
```diff
@@ -9,17 +9,17 @@
 //
 // 	http://www.apache.org/licenses/LICENSE-2.0
 //
-// Unless required by applicable law or agreed to in writing, software
+
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:12]: 🕵️

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 5]
```diff
@@ -146,16 +146,48 @@ mod worker_messages {
 	pub struct ImportJustification<B: BlockT>(pub Origin, pub B::Hash, pub NumberFor<B>, pub Justification);
 }
 
-struct BlockImportWorker<B: BlockT, Transaction> {
+/// The process of importing blocks.
+///
+/// This polls the `block_import` for new blocks to import and than awaits on importing these blocks.
```
Comment (andresilva) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:151]:
```suggestion
/// This polls the `block_import_receiver` for new blocks to import and than awaits on importing these blocks.
```

**Review by andresilva — APPROVED** [review]

File: primitives/consensus/common/src/import_queue/basic_queue.rs (hunk 2 above)
Comment (bkchr) [reply primitives/consensus/common/src/import_queue/basic_queue.rs:377]: In the end, after the first error has occurred, we would always return `Err(Canceled)` anyway.

File: primitives/consensus/common/src/import_queue/basic_queue.rs (hunk 4 above)
Comment (bkchr) [reply primitives/consensus/common/src/import_queue/basic_queue.rs:12]: Hmm xD

File: primitives/consensus/common/src/import_queue/basic_queue.rs [hunk 6]
```diff
@@ -146,16 +146,48 @@ mod worker_messages {
 	pub struct ImportJustification<B: BlockT>(pub Origin, pub B::Hash, pub NumberFor<B>, pub Justification);
 }
 
-struct BlockImportWorker<B: BlockT, Transaction> {
+/// The process of importing blocks.
+///
+/// This polls the `block_import_receiver` for new blocks to import and than awaits on importing these blocks.
+/// After each block is imported, this async function yields once to give other futures the possibility
```
Comment (octol) [inline primitives/consensus/common/src/import_queue/basic_queue.rs:152]: Nitpick: line length

(Not sure if we bother about this tbh?)

**Review by octol — APPROVED** [review]
Looks great!

File: primitives/consensus/common/src/import_queue/basic_queue.rs (hunk 6 above)
Comment (bkchr) [reply primitives/consensus/common/src/import_queue/basic_queue.rs:152]: CI is happy :P

**Pull request merged**